timesync/clock.go

```diff
@@ -1,211 +1,71 @@
-type LayerTimer chan types.LayerID
-
-type ManualClock struct {
-	*Ticker
-	genesisTime time.Time
-}
-
-func (t *ManualClock) Tick() {
-	t.notifyOnTick()
+type Clock interface {
+	Now() time.Time
 }
 
-func NewManualClock(genesisTime time.Time) *ManualClock {
-	t := &ManualClock{
-		Ticker: &Ticker{
-			nextLayerToTick: 1,
-			stop:            make(chan struct{}),
-			subscribers:     make(map[LayerTimer]struct{}),
-		},
-		genesisTime: genesisTime,
-	}
-	t.StartNotifying()
-	return t
-}
+type RealClock struct{}
 
-func (t ManualClock) GetGenesisTime() time.Time {
-	return t.genesisTime
+func (RealClock) Now() time.Time {
+	return time.Now()
 }
 
 type TimeClock struct {
 	*Ticker
 	tickInterval time.Duration
 	startEpoch   time.Time
-	time         Clock
+	stop         chan struct{}
+	once         sync.Once
+	log          log.Log
 }
 
-func NewTicker(time Clock, tickInterval time.Duration, startEpoch time.Time) *TimeClock {
+func NewClock(c Clock, tickInterval time.Duration, genesisTime time.Time, logger log.Log) *TimeClock {
 	t := &TimeClock{
-		Ticker: &Ticker{
-			nextLayerToTick: 1,
-			stop:            make(chan struct{}),
-			subscribers:     make(map[LayerTimer]struct{}),
-			layerChannels:   make(map[types.LayerID]chan struct{}),
-		},
+		Ticker:       NewTicker(c, LayerConv{duration: tickInterval, genesis: genesisTime}),
 		tickInterval: tickInterval,
-		startEpoch:   startEpoch,
-		time:         time,
+		startEpoch:   genesisTime,
+		stop:         make(chan struct{}),
+		once:         sync.Once{},
+		log:          logger,
 	}
-	t.init()
+	go t.startClock()
 	return t
 }
 
-func (t *TimeClock) init() {
-	var diff time.Duration
-	log.Info("start clock interval is %v", t.tickInterval)
-	if t.time.Now().Before(t.startEpoch) {
-		t.nextLayerToTick = 1
-		diff = t.startEpoch.Sub(t.time.Now())
-	} else {
-		t.updateLayerID()
-		diff = t.tickInterval - (t.time.Now().Sub(t.startEpoch) % t.tickInterval)
-	}
+func (t *TimeClock) startClock() {
+	t.log.Info("starting global clock now=%v genesis=%v", t.clock.Now(), t.startEpoch)
 
-	go t.startClock(diff)
-}
-
-func (t *TimeClock) updateLayerID() {
-	t.nextLayerToTick = types.LayerID((t.time.Now().Sub(t.startEpoch) / t.tickInterval) + 2)
-}
-
-func (t *TimeClock) startClock(diff time.Duration) {
-	log.Info("starting global clock now=%v genesis=%v", t.time.Now(), t.startEpoch)
-	log.Info("global clock going to sleep for %v", diff)
-
-	tmr := time.NewTimer(diff)
-	select {
-	case <-tmr.C:
-		break
-	case <-t.stop:
-		return
-	}
-	t.notifyOnTick()
-	tick := time.NewTicker(t.tickInterval)
-	log.Info("clock waiting on event, tick interval is %v", t.tickInterval)
 	for {
+		currLayer := t.Ticker.conv.TimeToLayer(t.clock.Now())    // get current layer
+		nextTickTime := t.Ticker.conv.LayerToTime(currLayer + 1) // get next tick time for the next layer
+		diff := nextTickTime.Sub(t.clock.Now())
+		tmr := time.NewTimer(diff)
```

Comment:
you are leaking timers here, all timers need to be stopped.

Reply:
To my understanding, there is no leak of whatever kind.
timer vs ticker: https://groups.google.com/forum/#!topic/golang-nuts/cCdm0Ixwi9A
Timers: https://medium.com/@oboturov/golang-time-after-is-not-garbage-collected-4cbc94740082

Note that I close in the case of receiving term signal.